source/manual/intro-to-docker.html.md

```diff
@@ -0,0 +1,350 @@
+
+It seems like our Ruby image isn’t doing quite what we need, so we’re going to create our own image based off it.
+
+#### Create ruby + node image
+
+ We are going to create our own Docker image based off of the ruby one, and add in other dependencies such as node.
+
+To create our own image, we need a Dockerfile. A Dockerfile normally starts with a `FROM [image]` which bases your new image off an existing image. We can then execute other commands by prefixing them with `RUN`. Each RUN command adds a new layer. You can think of it as each step creating a new image, but we only care about the final image to come out of the last step.
+Create a Dockerfile in the content-publisher project:
+
+> Note: there will likely already be a Dockerfile, so just remove it for the tutorial.
+
+*   Create a `Dockerfile` in the root of the [content-publisher] project:
+
+```docker
+FROM ruby:2.6.3
+
+RUN curl -sL https://deb.nodesource.com/setup_12.x | bash
+
+RUN apt-get install nodejs
+```
+
+*   Build our new image:
+
+```shell
+$mac docker build -t content-publisher
+```
+
+This creates an image based on the Dockerfile we just added.
+
+`-t` tags, or names it, as `content-publisher`.
+
+*   Now we can start a container with our new image:
+
+```shell
+$mac docker run -it --rm -v $(pwd):/app -v content-publisher-bundle:/usr/local/bundle content-publisher bash
+```
+
+Each time you change the Dockerfile, you need to remember to rebuild the docker image.
+
+If we run the tests again, this time we see another problem about no database. That's because we haven't got PostgreSQL installed.
+
+#### Use PostgreSQL image/container
+
+Docker containers are quick to start and it's possible for them to talk to each other, so let's set up a separate container for postgres. This approach also follows the Docker way, which is to have one container per process.
+
+*  In another terminal, if we run the following to start a new container based on the postgres image:
+
+```shell
+$mac docker run -it --rm postgres
+```
+
+* Now to see what containers you have, run:
+
+```shell
+$mac docker ps
+```
+
+And you should see (ignore any other containers you've got running):
+
+```shell
+$mac
+CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
+a7b87be3e1bc        postgres            "docker-entrypoint.s…"   21 minutes ago      Up 21 minutes       5432/tcp            jovial_matsumoto
+```
+
+* Find the ID for the postgres container, and with it run:
+
+```shell
+$mac docker inspect [container ID for postgres]
+```
+
+Grab the IP address from the output. We will use that IP address to point the content-publisher container to the one running postgres. Rails will use this to run its tests against.
+
+
+*   Set these environment variables:
```

Comment:
I wonder if we need to make it clear that these are run inside the content-publisher container?